File: csharp/lib/Property/ExposureProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Structure for the image of the camera exposure.
    /// </summary>
    [MessagePackObject]
    public class ExposureProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "exposure_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Exposure time : Auto.
        /// </summary>
        public static readonly int ExposureTimeAuto = 0;

        /// <summary>
        /// ISO Sensitivity : Auto.
        /// </summary>
        public static readonly int IsoSensitivityAuto = 0;

        /// <summary>
        /// Mode of exposure.
        /// </summary>
        [Key("mode")]
        public string Mode { get; set; } = string.Empty;

        /// <summary>
        /// Compensation value of EV.
        /// </summary>
        [Key("ev_compensation")]
        public float EvCompensation { get; set; }

        /// <summary>
        /// Time of exposure [usec].
        /// </summary>
        [Key("exposure_time")]
        public int ExposureTime { get; set; }

        /// <summary>
        /// ISO sensitivity. (100,200,400,800 ...)
        /// </summary>
        [Key("iso_sensitivity")]
        public int IsoSensitivity { get; set; }

        /// <summary>
        /// Exposure metering mode.
        /// </summary>
        [Key("metering")]
        public string Metering { get; set; } = string.Empty;

        /// <summary>
        /// Target region of the camera exposure.
        /// </summary>
        [Key("target_region")]
        public RectangleRegionParameter TargetRegion { get; set; } =
            new RectangleRegionParameter();

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return string.Format(
                "{0}(left={1} top={2} width={3} height={4})",
                this.Key, this.Mode, this.EvCompensation,
                this.ExposureTime, this.IsoSensitivity,
                this.Metering, this.TargetRegion);
        }
    }

    /// <summary>
    /// Structure for the region of plane for AE or ROI, etc.
    /// </summary>
    [MessagePackObject]
    public class RectangleRegionParameter
    {
        /// <summary>
        /// Upper position of region from origin.
        /// </summary>
        [Key("top")]
        public int Top { get; set; }

        /// <summary>
        /// Left  position of region from origin.
        /// </summary>
        [Key("left")]
        public int Left { get; set; }

        /// <summary>
        /// Bottom position of region from origin.
        /// </summary>
        [Key("bottom")]
        public int Bottom { get; set; }

        /// <summary>
        /// Right position of region from origin.
        /// </summary>
        [Key("right")]
        public int Right { get; set; }

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return string.Format(
                "Top={0} Left={1} Bottom={2} Right={3}",
                this.Top, this.Left, this.Bottom, this.Right);
        }
    }

    /// <summary>
    /// Exposure modes.
    /// </summary>
    public static partial class ExposureMode
    {
        /// <summary>
        /// Auto.
        /// </summary>
        public static readonly string Auto = "auto";

        /// <summary>
        /// Hold.
        /// </summary>
        public static readonly string Hold = "hold";

        /// <summary>
        /// Manual.
        /// </summary>
        public static readonly string Manual = "manual";

        /// <summary>
        /// GainFix.
        /// </summary>
        public static readonly string GainFix = "gainfix";

        /// <summary>
        /// TimeFix.
        /// </summary>
        public static readonly string TimeFix = "timefix";
    }

    /// <summary>
    /// Exposure metering modes.
    /// </summary>
    public static partial class ExposureMetering
    {
        /// <summary>
        /// None.
        /// </summary>
        public static readonly string None = "none";

        /// <summary>
        /// Average.
        /// </summary>
        public static readonly string Average = "average";

        /// <summary>
        /// Center weighted.
        /// </summary>
        public static readonly string CenterWeighted = "center_weighted";

        /// <summary>
        /// Spot.
        /// </summary>
        public static readonly string Spot = "spot";

        /// <summary>
        /// Matrix.
        /// </summary>
        public static readonly string Matrix = "matrix";
    }
}

File: java/senscord/src/senscord_properties/ExposureProperty.java
/*
 * SPDX-FileCopyrightText: 2019-2022 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Native;
import com.sun.jna.Structure;

import java.util.Arrays;
import java.util.List;

import senscord_types.RectangleRegionParameter;

/**
 * Structure for the image of the camera exposure.
 * 
 * @see "senscord::ExposureProperty"
 */
public class ExposureProperty extends Structure {
    public byte[] mode;
    public float ev_compensation;
    public int exposure_time;
    public int iso_sensitivity;
    public byte[] metering;
    public RectangleRegionParameter target_region;

    public static class ByReference extends ExposureProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("mode", "ev_compensation", "exposure_time", "iso_sensitivity",
                "metering", "target_region");
    }

    /**
     * Constructor.
     */
    public ExposureProperty() {
        super();
        mode = new byte[64];
        metering = new byte[64];
    }

    public static interface Mode {
        /**
         * Auto.
         */
        public static final String SENSCORD_EXPOSURE_MODE_AUTO = "auto";

        /**
         * Hold.
         */
        public static final String SENSCORD_EXPOSURE_MODE_HOLD = "hold";

        /**
         * Manual.
         */
        public static final String SENSCORD_EXPOSURE_MODE_MANUAL = "manual";

        /**
         * Gain fix.
         */
        public static final String SENSCORD_EXPOSURE_MODE_GAIN_FIX = "gainfix";

        /**
         * Time fix.
         */
        public static final String SENSCORD_EXPOSURE_MODE_TIME_FIX = "timefix";
    }

    public static interface Metering {
        /**
         * Mode.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_NONE = "note";

        /**
         * Average.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_AVERAGE = "average";

        /**
         * Center weighted.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_CENTER_WEIGHTED = "center_weighted";

        /**
         * Spot.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_SPOT = "spot";

        /**
         * Matrix.
         */
        public static final String SENSOCRD_EXPOSURE_METERING_MATRIX = "matrix";
    }

    /**
     * Get exposure mode from {@link Mode}.
     * 
     * @return exposure mode
     */
    public String GetMode() {
        return Native.toString(mode);
    }


    /**
     * Get EV compensation value.
     * 
     * @return EV compensation value.
     */
    public float GetEvCompensation() {
        return ev_compensation;
    }

    /**
     * Get exposure time.
     * 
     * @return exposure time
     */
    public int GetExposureTime() {
        return exposure_time;
    }

    /**
     * Get ISO sensitivity value.
     * 
     * @return ISO sensitivity value
     */
    public int GetIsoSensitivity() {
        return iso_sensitivity;
    }

    /**
     * Get exposure metering from {@link Metering}.
     * 
     * @return exposure metering metering
     */
    public String GetMetering() {
        return Native.toString(metering);
    }

    /**
     * Get rectangle region {@link RectangleRegionParameter}.
     * 
     * @return rectangle region
     */
    public RectangleRegionParameter GetTargetRegion() {
        return target_region;
    }

    /**
     * Set e exposure mode from {@link Mode}.
     * 
     * @param mode exposure mode
     */
    public void SetMode(String mode) {
        Arrays.fill(this.mode, (byte) 0);
        System.arraycopy(mode.getBytes(), 0, this.mode, 0, mode.length());
    }

    /**
     * Set EV compensation value.
     * 
     * @param ev_compensation EV compensation value
     */
    public void SetEvCompensation(float ev_compensation) {
        this.ev_compensation = ev_compensation;
    }

    /**
     * Set exposure time .
     * 
     * @param exposure_time exposure time
     */
    public void SetExposureTime(int exposure_time) {
        this.exposure_time = exposure_time;
    }

    /**
     * Set IOS sensitivity.
     * 
     * @param iso_sensitivity iso sensitivity
     */
    public void SetIsoSensitivity(int iso_sensitivity) {
        this.iso_sensitivity = iso_sensitivity;
    }

    /**
     * Set metering {@link Metering}.
     * 
     * @param metering exposure metering
     */
    public void SetMetering(String metering) {
        Arrays.fill(this.metering, (byte) 0);
        System.arraycopy(metering.getBytes(), 0, this.metering, 0, metering.length());
    }

    /**
     * Set rectangle region {@link RectangleRegionParameter}.
     * 
     * @param target_region rectangle region
     */
    public void SetTargetRegion(RectangleRegionParameter target_region) {
        this.target_region = target_region;
    }
}
